File: src/lib/code/mcs-216/section-1/session-5/3/3.py
def merge_sort(arr, left=0, right=None, depth=0):
    """
    Recursive Merge Sort Implementation
    :param arr: List to sort
    :param left: Left index
    :param right: Right index
    :param depth: Recursion depth for indentation
    :return: Sorted list
    """
    if right is None:
        right = len(arr) - 1

    # Print current state with indentation
    indent = "  " * depth
    print(f"{indent}mergeSort({left},{right}) → {arr[left:right+1]}")

    if left < right:
        # Calculate mid point (avoid overflow)
        mid = left + (right - left) // 2

        # Recursively sort first half
        merge_sort(arr, left, mid, depth + 1)

        # Recursively sort second half
        merge_sort(arr, mid + 1, right, depth + 1)

        # Merge the sorted halves
        merge(arr, left, mid, right, depth)
    else:
        print(f"{indent}  Base case: single element")

    return arr

def merge(arr, left, mid, right, depth):
    """
    Merge two sorted subarrays
    :param arr: Original array
    :param left: Start of left subarray
    :param mid: End of left subarray
    :param right: End of right subarray
    :param depth: Indentation level
    """
    indent = "  " * depth

    # Create temporary arrays
    left_arr = arr[left:mid + 1]
    right_arr = arr[mid + 1:right + 1]

    print(f"{indent}  Merge: Left={left_arr} Right={right_arr}")

    i = j = 0  # Initial indices for left and right subarrays
    k = left   # Initial index of merged subarray

    # Merge the temp arrays back
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            print(f"{indent}    Compare {left_arr[i]} <= {right_arr[j]} → Take {left_arr[i]}")
            i += 1
        else:
            arr[k] = right_arr[j]
            print(f"{indent}    Compare {left_arr[i]} > {right_arr[j]} → Take {right_arr[j]}")
            j += 1
        k += 1

    # Copy remaining elements
    while i < len(left_arr):
        arr[k] = left_arr[i]
        print(f"{indent}    Append remaining left: {left_arr[i]}")
        i += 1
        k += 1

    while j < len(right_arr):
        arr[k] = right_arr[j]
        print(f"{indent}    Append remaining right: {right_arr[j]}")
        j += 1
        k += 1

    print(f"{indent}  Result: {arr[left:right+1]}\n")

# Driver code
if __name__ == "__main__":
    arr = [200, 150, 50, 100, 75, 25, 10, 5]

    print("=== Merge Sort Implementation in Python ===")
    print(f"Original array: {arr}\n")

    merge_sort(arr)

    print(f"Sorted array: {arr}")

File: src/lib/code/mcs-216/section-1/session-5/3/3.rs
/// Merge Sort Implementation in Rust
/// Time Complexity: O(n log n)
/// Space Complexity: O(n)

/// Merge two sorted subarrays
///
/// # Arguments
/// * `arr` - Mutable slice to sort
/// * `left` - Starting index of left subarray
/// * `mid` - Ending index of left subarray
/// * `right` - Ending index of right subarray
/// * `depth` - Recursion depth for indentation
fn merge(arr: &mut [i32], left: usize, mid: usize, right: usize, depth: usize) {
    let indent = "  ".repeat(depth);

    // Create temporary vectors
    let left_arr: Vec<i32> = arr[left..=mid].to_vec();
    let right_arr: Vec<i32> = arr[mid + 1..=right].to_vec();

    println!(
        "{}  Merge: Left={:?} Right={:?}",
        indent, left_arr, right_arr
    );

    let mut i = 0; // Index for left_arr
    let mut j = 0; // Index for right_arr
    let mut k = left; // Index for merged array

    // Merge the temp arrays
    while i < left_arr.len() && j < right_arr.len() {
        if left_arr[i] <= right_arr[j] {
            arr[k] = left_arr[i];
            println!(
                "{}    Compare {} <= {} → Take {}",
                indent, left_arr[i], right_arr[j], left_arr[i]
            );
            i += 1;
        } else {
            arr[k] = right_arr[j];
            println!(
                "{}    Compare {} > {} → Take {}",
                indent, left_arr[i], right_arr[j], right_arr[j]
            );
            j += 1;
        }
        k += 1;
    }

    // Copy remaining elements
    while i < left_arr.len() {
        arr[k] = left_arr[i];
        println!("{}    Append remaining left: {}", indent, left_arr[i]);
        i += 1;
        k += 1;
    }

    while j < right_arr.len() {
        arr[k] = right_arr[j];
        println!("{}    Append remaining right: {}", indent, right_arr[j]);
        j += 1;
        k += 1;
    }

    println!("{}  Result: {:?}\n", indent, &arr[left..=right]);
}

/// Recursive merge sort function
///
/// # Arguments
/// * `arr` - Mutable slice to sort
/// * `left` - Left index
/// * `right` - Right index
/// * `depth` - Recursion depth for indentation
fn merge_sort(arr: &mut [i32], left: usize, right: usize, depth: usize) {
    let indent = "  ".repeat(depth);

    // Print current subarray
    print!(
        "{}mergeSort({},{}) → {:?}\n",
        indent,
        left,
        right,
        &arr[left..=right]
    );

    if left < right {
        // Calculate mid point (avoid overflow)
        let mid = left + (right - left) / 2;

        // Sort first half
        merge_sort(arr, left, mid, depth + 1);

        // Sort second half
        merge_sort(arr, mid + 1, right, depth + 1);

        // Merge sorted halves
        merge(arr, left, mid, right, depth);
    } else {
        println!("{}  Base case: single element", indent);
    }
}

fn main() {
    let mut arr = [200, 150, 50, 100, 75, 25, 10, 5];
    let n = arr.len();

    println!("=== Merge Sort Implementation in Rust ===");
    println!("Original array: {:?}\n", arr);

    merge_sort(&mut arr, 0, n - 1, 0);

    println!("Sorted array: {:?}", arr);
}
